### PR TITLE
Blobs migration update

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Finishes up the migration guide from Microsoft.Azure.Storage.Blobs to Azure.Storage.Blobs. This PR adds a couple extra sections, fixes some inconsistencies, and makes samples better mirror each other.

This PR contains many tiny changes which touch nearly the entire document. They are categorized into their own commits. It may be easier to review individual commits, each with their own commit descriptions.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md [hunk 1]
```diff
@@ -55,97 +56,46 @@ dotnet add package Azure.Storage.Blobs
...
-    BlobName = blobName,
-    ExpiresOn = new DateTimeOffset
-    IPRange = new SasIPRange(System.Net.IPAddress.None, System.Net.IPAddress.None)
-};
-// Set Permissions
-sasBuilder.SetPermissions(BlobSasPermissions.Read);
 
-// Create SasQueryParameters
-BlobSasQueryParameters parameters = sasBuilder.ToSasQueryParameters(sharedKeyCredential);
+The new library only supports constructing a client with a fully constructed SAS URI. Note that since client URIs are immutable once created, there is currently no way to rotate a SAS in a client.
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:94]: I'd rather say that "in order to rotate SAS new instance must be created" or something like that. We'd rather point user to alternative than say it's impossible.

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md [hunk 2]
```diff
@@ -247,24 +239,38 @@ async static Task CreateStoredAccessPolicyAsync(string containerName)
 }
 ```
 
-### Client hierarchy
+### Client structure
 
-In the interest of simplifying the API surface we've made a three top level clients that can be used to interact with a majority of your resources: `BlobServiceClient`, `BlobContainerClient`, and `BlobClient`.
+The legacy SDK was structured like an object model. There were container and blob objects that held state and required the user to manually call their update methods. But blob contents were not a part of this state and had to be uploaded/downloaded whenever they were to be interacted with. This became increasingly confusing over time, and increasingly susceptible to thread safety issues.
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:244]: [Object model](https://en.wikipedia.org/wiki/Object_model) isn't synonym for state-full. Both `BlobClient` and `CloudBlob` together with surround data models are object oriented representations of blobs. The difference is that one is state-full and the other one is state-less and this is the terminology that should be used to describe the difference.

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md [hunk 3]
```diff
@@ -247,24 +239,38 @@ async static Task CreateStoredAccessPolicyAsync(string containerName)
...
 
-### Client constructors
+The hierarchical structure of Azure Blob Storage can be understood by the following diagram:
+![Blob Storage Hierarchy](https://docs.microsoft.com/en-us/azure/storage/blobs/media/storage-blobs-introduction/blob1.png)
+
+In the interest of simplifying the API surface, v12 uses three top level clients to match this structure that can be used to interact with a majority of your resources: `BlobServiceClient`, `BlobContainerClient`, and `BlobClient`. Note that blob-type-specific operations can still be accessed by their specific clients, as in v11.
+
+`BlobClient` is NOT a replacement of v11's `CloudBlob`, which served as a base class for the three blob type classes. v12 contains `BlobBaseClient` as a replacement base class. `BlobClient` is a SDK-only concept that trades off advanced functionality like partial blob updates for an easier to understand abstraction of blobs, wher you do not need to worry about blob types or their implementation mechanisms. They are backed up by block blobs in Storage, and you can switch your code over to using `BlockBlobClient`s with no extra steps involving your already-stored data.
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:253]: typo `wher`

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md [hunk 4]
```diff
@@ -247,24 +239,38 @@ async static Task CreateStoredAccessPolicyAsync(string containerName)
...
 
 | v11 | v12 |
 |-------|--------|
-| `CloudStorageAccount` | `BlobServiceClient` |
+| `CloudBlobClient` | `BlobServiceClient` |
 | `CloudBlobContainer`  | `BlobContainerClient` |
-| `CloudBlobDirectory` | Not supported |
+| `CloudBlobDirectory` | No equivalent |
+| No equivalent | `BlobClient`
 | `CloudBlob` | `BlobBaseClient` |
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:267]: I'd rather rephrase this section in such a way that user is encouraged to try `BlobClient` first. Suggesting that there's no equivalent and saying that BlobClient is not a replacement for CloudBlob above might discourage from trying it first.

Maybe we should say that if they don't have any peculiar requirements then they should be trying BlobClient first?

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md [hunk 5]
```diff
@@ -317,58 +323,221 @@ BlobClient blobClient = containerClient.GetBlobClient(blobName);
...
+{
+    // process page
+
+    foreach (BlobItem item in page.Values)
+    {
+        // process blob listing
+    }
+
+    // access continuation token if desired
+    string continuationToken = page.ContinuationToken;
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:432]: might be worth adding sample how to start iteration from continuation token. I had few customers asking for it as they couldn't discover that.

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md [hunk 6]
```diff
@@ -317,58 +323,221 @@ BlobClient blobClient = containerClient.GetBlobClient(blobName);
...
+You could also make a SAS using a predefined policy stored on the service, instead of defining it in code.
+
+```csharp
+string policyId; // the id of the stored policy
+sasBlobToken = blob.GetSharedAccessSignature(null, policyName);
+```
+
+v12
+
+The modern SDK uses a builder pattern for constructing a SAS token. Clients are not involved in the process.
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:507]: we should add @amnguye 's work here - how to generate SAS using blobClient. it's going out pretty soon.

**Review by kasobol-msft — COMMENTED** [review]
Overall looking good.

**Comment by kasobol-msft** [comment]
Btw. for V12 samples - could we move them to https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/storage/Azure.Storage.Blobs/samples ? and include using #region thing? so that compiler makes sure they don't get out of date.

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md (hunk 5 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:432]:
```csharp
// set this to already existing continuation token to pick up where you previously left off
string initialContinuationToken = null;
```
That's the intent of these lines at the start of this sample. Perhaps we need to better highlight this section? Or rephrase?

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:432]: no it's good. I'm blind. thanks!

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md (hunk 6 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md:507]: I've spoken with her about adding this when it becomes available. That probably won't make it into this PR.

File: sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md (hunk 4 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md]: I added two subsections to this section. One of them is "Migrating from CloudBlockBlob", who's opening paragraph is "We recommend `BlobClient` as a starting place when migrating code that used v11's `CloudBlockBlob`."

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs/AzureStorageNetMigrationV12.md]: looks good.

**Review by kasobol-msft — APPROVED** [review]

**Pull request merged**